File: src/main/java/com/github/nigelzor/hogs/Tile.java
package com.github.nigelzor.hogs;

import java.util.Set;

import com.google.common.base.Objects;
import com.google.common.base.Optional;

public class Tile implements Position {
	private Set<Direction> connections;
	private Optional<Objective> objective;
	private Set<Player> players;

	public Tile(Set<Direction> connections) {
		this.connections = connections;
		this.objective = Optional.absent();
	}

	public Tile(Set<Direction> connections, Objective objective) {
		this.connections = connections;
		this.objective = Optional.of(objective);
	}

	private Tile(Set<Direction> connections, Optional<Objective> objective, Set<Player> players) {
		this.connections = connections;
		this.objective = objective;
		this.players = players;
	}

	public Tile rotate(Rotation rotation) {
		return new Tile(Direction.rotate(connections, rotation), objective, players);
	}

	public Set<Direction> getConnections() {
		return connections;
	}

	public boolean connectsTo(Direction direction) {
		return connections.contains(direction);
	}

	@Override
	public Set<Player> getPlayers() {
		return players;
	}

	public Optional<Objective> getObjective() {
		return objective;
	}

	@Override
	public int hashCode() {
		return Objects.hashCode(connections, objective, players);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) return true;
		if (obj == null) return false;
		if (getClass() != obj.getClass()) return false;
		Tile other = (Tile) obj;
		return Objects.equal(connections, other.connections)
				&& Objects.equal(objective, other.objective)
				&& Objects.equal(players, other.players);
	}
}

File: src/main/kotlin/com/github/nigelzor/hogs/Tile.kt
package com.github.nigelzor.hogs;

import java.util.HashSet

import kotlin.Int
import kotlin.Int as BTile

data class Tile(val connections: BTile, val objective: Objective? = null, override val players: MutableSet<Int> = HashSet()): Position {
	companion object {
		fun fromDirections(vararg ds : Direction): Tile {
			return Tile(BTiles.fromDirections(ds.toList()))
		}
	}

	public fun rotate(rotation: Rotation): Tile {
		return Tile(BTiles.rotate(connections, rotation), objective, players)
	}

	public fun connectsTo(direction: Direction): Boolean {
		return BTiles.contains(connections, direction)
	}

	public fun clone(): Tile {
		return copy(players=HashSet(players))
	}

}
